packages/client/internals/MonacoOutput.vue

```diff
@@ -0,0 +1,167 @@
+<script setup lang="ts">
+import { debounce } from '@antfu/utils'
+import { useVModel } from '@vueuse/core'
+import { computed, ref, shallowRef, watch } from 'vue'
+import { isDark } from '../logic/dark'
+import { isPrintMode } from '../logic/nav'
+import type { JavaScriptExecutionLog } from '../logic/runCode'
+import { runJavaScript } from '../logic/runCode'
+import { useSlideContext } from '../context'
+import IconButton from './IconButton.vue'
+
+const props = withDefaults(
+  defineProps<{
+    modelValue: string
+    lang: string
+    autorun?: boolean | 'once'
+    height?: string
+    highlightOutput?: boolean
+  }>(),
+  {
+    highlightOutput: true,
+  },
+)
+
+const emit = defineEmits(['update:modelValue'])
+const code = useVModel(props, 'modelValue', emit)
+
+const { $renderContext } = useSlideContext()
+const disabled = computed(() => !['slide', 'presenter'].includes($renderContext.value))
+
+const autorun = isPrintMode.value ? 'once' : props.autorun
+const isRunning = ref(autorun)
+const output = shallowRef<JavaScriptExecutionLog[]>()
+
+let tsModule: typeof import('typescript') | undefined
+
+let highlight: ((code: string) => string) | undefined
+
+const run = debounce(200, async () => {
+  if (disabled.value)
+    return
+
+  const setAsRunning = setTimeout(() => {
+    isRunning.value = true
+  }, 500)
+
+  if (!highlight && props.highlightOutput) {
+    const { shiki, themes } = await import('#slidev/shiki')
+    const highlighter = await shiki
+    highlight = (code: string) => highlighter.codeToHtml(code, {
+      lang: 'javascript',
+      theme: typeof themes === 'string'
+        ? themes
+        : isDark.value
+          ? themes.dark || 'vitesse-dark'
+          : themes.light || 'vitesse-light',
+    })
+  }
+
+  const { transpile } = (tsModule ??= await import('typescript'))
+  const js = props.lang === 'typescript'
+    ? transpile(code.value, {
+      module: tsModule.ModuleKind.ESNext,
+    })
+    : code.value
```

Comment:
We might want to throw a warning for non ts/js lang

Reply:
Added a `setup/runners.ts` to register custom runners.